library/md5.c

```diff
@@ -302,29 +318,46 @@ void mbedtls_md5_finish( mbedtls_md5_context *ctx, unsigned char output[16] )
     last = ctx->total[0] & 0x3F;
     padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );
 
-    mbedtls_md5_update( ctx, md5_padding, padn );
-    mbedtls_md5_update( ctx, msglen, 8 );
+    if( ( ret = mbedtls_md5_update_ext( ctx, md5_padding, padn ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_update_ext( ctx, msglen, 8 ) ) != 0 )
+            return( ret );
 
     PUT_UINT32_LE( ctx->state[0], output,  0 );
     PUT_UINT32_LE( ctx->state[1], output,  4 );
     PUT_UINT32_LE( ctx->state[2], output,  8 );
     PUT_UINT32_LE( ctx->state[3], output, 12 );
+
+    return( 0 );
 }
 
 #endif /* !MBEDTLS_MD5_ALT */
 
 /*
  * output = MD5( input buffer )
  */
-void mbedtls_md5( const unsigned char *input, size_t ilen, unsigned char output[16] )
+int mbedtls_md5_ext( const unsigned char *input,
+                     size_t ilen,
+                     unsigned char output[16] )
 {
+    int ret;
     mbedtls_md5_context ctx;
 
     mbedtls_md5_init( &ctx );
-    mbedtls_md5_starts( &ctx );
-    mbedtls_md5_update( &ctx, input, ilen );
-    mbedtls_md5_finish( &ctx, output );
+
+    if( ( ret = mbedtls_md5_starts_ext( &ctx ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_update_ext( &ctx, input, ilen ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_finish_ext( &ctx, output ) ) != 0 )
+            return( ret );
+
     mbedtls_md5_free( &ctx );
+
+    return( 0 );
```

Comment:
I am wondering whether we should make the call to `mbedtls_XXX_free` a "cleanup" step that we're always jumping to when `starts/update/finish` fail. To decide this, I think we should expand on the documentation of the latter functions. For example, does a failure mean that the context must not be used at all anymore, or is any alternative implementation responsible for making `mbedtls_XXX_free` safe to be called even if some intermediate operation failed?